### PR TITLE
Topic/query param key

### PR DESCRIPTION
**Pull request by bryce-anderson**
this PR adds a typeclass that will represent a query key. It may not be finished (maybe it is), but I wanted to get this up for feedback. 

### Changes
* ```QueryParamEncoder[T].encode(T) : Seq[QueryParameterValue]``` instead of a single value. This is needed for the following point to work.
* Varadiac methods were dropped in favor of a single value which also is a type class. Multiple values can be handled by a typeclass that takes a ```Seq[T]```; typeclasses for other types of collections are needed.
* Quite a few methods were dropped from the ```Uri``` class, but not as many as I'd hoped. I cant help but feel the ```Maybe``` and ```Option``` receiving methods are largely clutter: the same functionality can be done with ```cata``` and ```fold``` or other standard techniques. Perhaps I just haven't needed them yet and they are really awsome: its really a matter of convenience.

### CODE REVIEW

File: core/src/main/scala/org/http4s/QueryParam.scala [hunk 1]
```diff
@@ -26,13 +26,24 @@ object QueryParam {
...
 }
 
+trait QueryParamKeyLike[T] {
+  def getKey(t: T): QueryParameterKey
+}
+
+object QueryParamKeyLike {
+  def apply[T](implicit ev: QueryParamKeyLike[T]): QueryParamKeyLike[T] = ev
+
+  implicit val stringKey: QueryParamKeyLike[String] = new QueryParamKeyLike[String] {
```
Comment (bryce-anderson) [inline core/src/main/scala/org/http4s/QueryParam.scala]: I'm not sure I understand: do you mean an instance of ```QueryParamKeyLike[QueryParameterKey]```?

Comment (julien-truffaut) [inline core/src/main/scala/org/http4s/QueryParam.scala]: yes

File: core/src/test/scala/org/http4s/QueryParamCodecLaws.scala [hunk 2]
```diff
@@ -15,7 +15,7 @@ object QueryParamCodecLaws {
   def apply[T: Arbitrary: Equal: QueryParamDecoder: QueryParamEncoder] = new Properties("QueryParamCodec") {
 
     property("decode . encode == successNel") = forAll { value: T =>
-      (QueryParamDecoder[T].decode _ compose QueryParamEncoder[T].encode)(value) === value.successNel
+      QueryParamEncoder[T].encode(value).map(QueryParamDecoder[T].decode).head === value.successNel
```
Comment (julien-truffaut) [inline core/src/test/scala/org/http4s/QueryParamCodecLaws.scala]: this test should fail for all `QueryParamEncoder` with more than one value

Comment (bryce-anderson) [inline core/src/test/scala/org/http4s/QueryParamCodecLaws.scala]: Ah, good point.

**Comment by bryce-anderson** [comment]
Where do we stand on this? I know @julien-truffaut would like to do a little cleanup based on this involving removing the internal ```_``` prefixed methods, but I'm not entirely sure what he wants so I'm keen to leave that to him. 
I'd like to get this in before it gets too stale, so I'd appreciate suggestions or the OK to merge it. After that, we should decide if we are going to change the representation from a ```String``` (I think this is a good idea, maybe we can preserve laziness like we did with headers, maybe its not an issue and we just eat a few CPU cycles).

**Comment by rossabaker** [comment]
I haven't followed this very closely, but it looks reasonable.

I still like the LinearSeqOptimized query string approach.

**Comment by before** [comment]
I like it too.

**Comment by bryce-anderson** [comment]
Cool, if @julien-truffaut doesn't object in the next while, I'll merge it. @rossabaker, by ```LinearSeqOptimized``` I assume you are referring to the form of the query after parsing the ```String```.

**Comment by rossabaker** [comment]
Yes.

**Comment by julien-truffaut** [comment]
I think there is something wrong on `QueryParamEncoder` / `QueryParamDecoder` type class because this change makes them non "symmetrical". `QueryParamEncoder` produces a `Seq[QueryParamValue]` but `QueryParamDecoder` accepts a single `QueryParamValue`. I don't have an answer but I think we should take some time to think about the law of these 2 type classes.

**Comment by bryce-anderson** [comment]
@julien-truffaut, the break in symmetry comes from using nested typeclasses to encode list type structures. If we Make all methods take a ```Seq[T]```, I think it would restore symmetry at the cost of requiring all parameters be a ```Seq```. On the plus side, the special case of ```Unit``` for when we don't want any params is dropped. I haven't attempted to see if that works, I'll give it a shot.

**Comment by bryce-anderson** [comment]
Well, the zero param case is still a problem because it cannot infer the type of a ```Nil``` which is needed for the typeclass, so its free to choose any, and thus has ambiguity problems. This could be cleared up by renaming a class of the ```+?``` methods, much like is done with the ```Option``` and ```Maybe``` versions. I'm going to rename a few of them and see how it goes.

**Comment by bryce-anderson** [comment]
@julien-truffaut, that last commit restores the symmetry which I think is a good thing. Sadly, a couple of the ```+?``` methods had to be renamed to ```+*?``` to avoid ambiguities. I think this is acceptable, and maybe there is a better name?

**Comment by bryce-anderson** [comment]
I was happy enough with this not to wait any longer to merge it so I can move on to other things. If there are other changes to be made (almost certainly are) lets do another PR.

**Comment by rossabaker** [comment]
+1. Any further bikeshedding of the query can be done when it goes
LinearSeqOptimized.

**Comment by bryce-anderson** [comment]
I'm working on that now.

**Comment by julien-truffaut** [comment]
I don't think that the case with zero value is very common, so I believe it could be acceptable to use `Seq.empty[String]` or `Seq.empty[Unit]`.

**Pull request merged**